smithy-model/src/main/java/software/amazon/smithy/model/traits/HttpChecksumProperties.java

```diff
@@ -0,0 +1,227 @@
+        return new Builder();
+    }
+
+    /**
+     * Create a {@code HttpChecksumProperties} from {@link Node}.
+     *
+     * @param node {@code Node} to create the {@code HttpChecksumProperties}
+     * @return Returns the created {@code HttpChecksumProperties}.
+     * @throws ExpectationNotMetException if the given {@code node} is invalid.
+     */
+    public static HttpChecksumProperties fromNode(Node node) {
+        ObjectNode value = node.expectObjectNode().warnIfAdditionalProperties(KEYS);
+        HttpChecksumProperties.Builder builder = builder();
+        value.getStringMember(PREFIX).map(StringNode::getValue).ifPresent(builder::prefix);
+        value.getArrayMember(ALGORITHMS).ifPresent(algorithms ->
+                Node.loadArrayOfString(ALGORITHMS, algorithms).forEach(builder::addAlgorithm));
+        value.getArrayMember(LOCATIONS).ifPresent(locationNodes -> {
+            builder.clearLocations();
+            locationNodes.forEach(locationNode -> builder.addLocation(Location.fromNode(locationNode)));
+        });
+        return builder.build();
+    }
+
+    /**
+     * @return set of locations supported within the HttpChecksum trait.
+     */
+    public Set<Location> getLocations() {
+        return locations;
+    }
+
+    /**
+     * @return prefix to be used with checksum algorithm to build header or trailer name.
+     */
+    public String getPrefix() {
+        return prefix;
+    }
+
+    /**
+     * @return set of string denoting the supported checksum algorithms.
+     */
+    public Set<String> getAlgorithms() {
+        return algorithms;
+    }
+
+    /**
+     * Take this object and create a builder that contains all of the
+     * current property values of this object.
+     *
+     * @return a builder for type HttpChecksumProperties
+     */
+    @Override
+    public SmithyBuilder<HttpChecksumProperties> toBuilder() {
+        return builder()
+                .prefix(getPrefix())
+                .algorithms(getAlgorithms())
+                .locations(getLocations());
+    }
+
+    /**
+     * Converts a value to a {@link Node}.
+     *
+     * @return Returns the creates Node.
+     */
+    @Override
+    public Node toNode() {
+        List<Node> locationNodes = locations.stream().map(Location::toNode).collect(Collectors.toList());
+        ObjectNode.Builder builder = Node.objectNodeBuilder()
+                .withMember(PREFIX, getPrefix())
+                .withMember(ALGORITHMS, Node.fromStrings(algorithms))
+                .withMember(LOCATIONS, Node.fromNodes(locationNodes));
+
+        return builder.build();
+    }
+
+    public static final class Builder implements SmithyBuilder<HttpChecksumProperties> {
+        private String prefix;
+        private Set<String> algorithms = new LinkedHashSet<>();
+        private Set<Location> locations = new LinkedHashSet<>();
+
+        private Builder() {
+            // Add "header" as default supported location.
```

Comment:
This should be done in the call to `build()` if the `locations` list is empty instead. Otherwise, it's possible for someone to create a builder, call `addLocation(Location.TRAILER)`, then `build()` and receive checksum properties that also contain `HEADER`.

Reply:
Thanks for catching this. I updated to set default location in build() if locations is empty. I  added a little extra logic to make sure we do not silently add "Header" in locations list, if service accidentally configures an empty list for locations property on the Trait. (We throw validation error if that happens).